### PR TITLE
[CB-8188] Add Medium Duty HA Entitlement enforcement

### PR DESCRIPTION
**Pull request by atxmemo**
Added medium duty ha entitlement capabilities and tests to SDX cluster creation logic

Implementation inspired by: https://github.com/hortonworks/cloudbreak/commit/a04709441071729e4019c2c487f2f7f0c9ee5fd9 

Flow is as follows:

1. Request comes into cloudbreak via SdxClusterRequest.java with enableMediumDutyHA attribute 
2. Logic in SdxService.java validates that the entitlement is enabled via the EntitlementService.java and that the runtime is greater or equal to 7.2.2 if enableMediumDutyHA is requested - errors out otherwise

Tests added validate the following scenarios:

- Valid request, valid runtime, valid entitlement permissions
- Valid request, valid runtime, invalid entitlement permissions
- Valid request, invalid (7.1.0) runtime, valid entitlement permissions
- Valid request, invalid (7.2.0) runtime, valid entitlement permissions

### CODE REVIEW

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterResponse.java [hunk 1]
```diff
@@ -37,14 +37,17 @@
 
     private boolean rangerRazEnabled;
 
+    private boolean mediumDutyHAEnabled;
```
Comment (orngejaket) [inline datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterResponse.java:40]: This is not needed.

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterDetailResponse.java [hunk 2]
```diff
@@ -18,7 +18,8 @@ public SdxClusterDetailResponse(SdxClusterResponse sdxClusterResponse, StackV4Re
                 sdxClusterResponse.getEnvironmentCrn(), sdxClusterResponse.getStackCrn(),
                 sdxClusterResponse.getClusterShape(), sdxClusterResponse.getCloudStorageBaseLocation(),
                 sdxClusterResponse.getCloudStorageFileSystemType(), sdxClusterResponse.getRuntime(),
-                sdxClusterResponse.getRangerRazEnabled());
+                sdxClusterResponse.getRangerRazEnabled(),
```
Comment (orngejaket) [inline datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterDetailResponse.java:21]: Yea this doesn't need to be part of the response.  Remove it for the other files.

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterRequest.java [hunk 3]
```diff
@@ -108,4 +110,12 @@ public boolean isEnableRangerRaz() {
     public void setEnableRangerRaz(boolean enableRangerRaz) {
         this.enableRangerRaz = enableRangerRaz;
     }
+
+    public boolean isEnableMediumDutyHA() {
```
Comment (kmanamcheri) [inline datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterRequest.java:114]: I don't understand why we need this? Shouldn't you just check if the cluster shape is Medium Duty?

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java [hunk 4]
```diff
@@ -452,6 +453,24 @@ private void validateRazEnablement(SdxClusterRequest sdxClusterRequest, Detailed
         }
     }
 
+    private void validateMediumDutyHaEnablement(SdxClusterRequest sdxClusterRequest, DetailedEnvironmentResponse environment) {
+        ValidationResultBuilder validationBuilder = new ValidationResultBuilder();
+        if (sdxClusterRequest.isEnableMediumDutyHA()) {
```
Comment (kmanamcheri) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java:458]: Shouldn't this check if the cluster shape requested is medium duty? If it is, then you validate that you have the entitlement.

File: datalake/src/test/java/com/sequenceiq/datalake/service/sdx/SdxServiceTest.java [hunk 5]
```diff
@@ -624,6 +625,100 @@ void testSdxCreateRazEnabled720Runtime() {
...
+        when(sdxClusterRepository.findByAccountIdAndEnvNameAndDeletedIsNull(anyString(), anyString())).thenReturn(new ArrayList<>());
+        long id = 10L;
+        when(sdxClusterRepository.save(any(SdxCluster.class))).thenAnswer(invocation -> {
+            SdxCluster sdxWithId = invocation.getArgument(0, SdxCluster.class);
+            sdxWithId.setId(id);
+            return sdxWithId;
+        });
+        when(clock.getCurrentTimeMillis()).thenReturn(1L);
+        mockEnvironmentCall(sdxClusterRequest, CloudPlatform.AWS);
+        sdxClusterRequest.setEnableMediumDutyHA(true);
```
Comment (orngejaket) [inline datalake/src/test/java/com/sequenceiq/datalake/service/sdx/SdxServiceTest.java:651]: remove.

File: datalake/src/test/java/com/sequenceiq/datalake/service/sdx/SdxServiceTest.java [hunk 6]
```diff
@@ -624,6 +625,100 @@ void testSdxCreateRazEnabled720Runtime() {
...
+        sdxClusterRequest.setRuntime(runtime);
+        sdxClusterRequest.setClusterShape(MEDIUM_DUTY_HA);
+        Map<String, String> tags = new HashMap<>();
+        tags.put("mytag", "tagecske");
+        sdxClusterRequest.addTags(tags);
+        sdxClusterRequest.setEnvironment("envir");
+        when(sdxClusterRepository.findByAccountIdAndEnvNameAndDeletedIsNull(anyString(), anyString())).thenReturn(new ArrayList<>());
+        long id = 10L;
+        mockEnvironmentCall(sdxClusterRequest, CloudPlatform.AWS);
+        sdxClusterRequest.setEnableMediumDutyHA(true);
```
Comment (orngejaket) [inline datalake/src/test/java/com/sequenceiq/datalake/service/sdx/SdxServiceTest.java:675]: remove

File: datalake/src/test/java/com/sequenceiq/datalake/service/sdx/SdxServiceTest.java [hunk 7]
```diff
@@ -624,6 +625,100 @@ void testSdxCreateRazEnabled720Runtime() {
...
+        sdxClusterRequest.setRuntime(invalidRuntime);
+        sdxClusterRequest.setClusterShape(MEDIUM_DUTY_HA);
+        Map<String, String> tags = new HashMap<>();
+        tags.put("mytag", "tagecske");
+        sdxClusterRequest.addTags(tags);
+        sdxClusterRequest.setEnvironment("envir");
+        when(sdxClusterRepository.findByAccountIdAndEnvNameAndDeletedIsNull(anyString(), anyString())).thenReturn(new ArrayList<>());
+        long id = 10L;
+        mockEnvironmentCall(sdxClusterRequest, CloudPlatform.AWS);
+        sdxClusterRequest.setEnableMediumDutyHA(true);
```
Comment (orngejaket) [inline datalake/src/test/java/com/sequenceiq/datalake/service/sdx/SdxServiceTest.java:715]: remove

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterRequest.java (hunk 3 above)
Comment (atxmemo) [reply datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterRequest.java]: yeah, good point.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java [hunk 8]
```diff
@@ -452,6 +453,24 @@ private void validateRazEnablement(SdxClusterRequest sdxClusterRequest, Detailed
         }
     }
 
+    private void validateMediumDutyHaEnablement(SdxClusterRequest sdxClusterRequest, DetailedEnvironmentResponse environment) {
+        ValidationResultBuilder validationBuilder = new ValidationResultBuilder();
+        if (sdxClusterRequest.getClusterShape().equals(SdxClusterShape.MEDIUM_DUTY_HA)) {
```
Comment (orngejaket) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java:458]: Swap this around, cluster shape could be null and thus null.equals throws a npe.

File: auth-connector/src/main/java/com/sequenceiq/cloudbreak/auth/altus/EntitlementService.java [hunk 9]
```diff
@@ -41,6 +41,9 @@
     @VisibleForTesting
     static final String CDP_RAZ = "CDP_RAZ";
 
+    @VisibleForTesting
+    static final String CDP_MEDIUM_DUTY_HA = "CDP_MEDIUM_DUTY_HA";
```
Comment (doktoric) [inline auth-connector/src/main/java/com/sequenceiq/cloudbreak/auth/altus/EntitlementService.java:45]: CDP_MEDIUM_DUTY_SDX ?

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java [hunk 10]
```diff
@@ -452,6 +453,24 @@ private void validateRazEnablement(SdxClusterRequest sdxClusterRequest, Detailed
...
+    private void validateMediumDutySdxEnablement(SdxClusterRequest sdxClusterRequest, DetailedEnvironmentResponse environment) {
+        ValidationResultBuilder validationBuilder = new ValidationResultBuilder();
+        if (SdxClusterShape.MEDIUM_DUTY_HA.equals(sdxClusterRequest.getClusterShape())) {
+            boolean mediumDutySdxEntitlementEnabled = entitlementService.mediumDutySdxEnabled(environment.getCreator(),
+                    Crn.safeFromString(environment.getCreator()).getAccountId());
+            if (!mediumDutySdxEntitlementEnabled) {
+                validationBuilder.error("Provisioning a Medium Duty SDX shape is not enabled for this account.");
+            }
+            if (!isMediumDutySdxSupported(sdxClusterRequest.getRuntime())) {
+                validationBuilder.error("Provisioning a Medium Duty SDX shape is only valid for CM version > 7.2.2 and not " + sdxClusterRequest.getRuntime());
```
Comment (kmanamcheri) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java:465]: Is 7.2.2 supported? If so, then it should be ">=" not ">".

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java [hunk 11]
```diff
@@ -452,6 +453,24 @@ private void validateRazEnablement(SdxClusterRequest sdxClusterRequest, Detailed
         }
     }
 
+    private void validateMediumDutySdxEnablement(SdxClusterRequest sdxClusterRequest, DetailedEnvironmentResponse environment) {
+        ValidationResultBuilder validationBuilder = new ValidationResultBuilder();
+        if (SdxClusterShape.MEDIUM_DUTY_HA.equals(sdxClusterRequest.getClusterShape())) {
+            boolean mediumDutySdxEntitlementEnabled = entitlementService.mediumDutySdxEnabled(environment.getCreator(),
+                    Crn.safeFromString(environment.getCreator()).getAccountId());
+            if (!mediumDutySdxEntitlementEnabled) {
+                validationBuilder.error("Provisioning a Medium Duty SDX shape is not enabled for this account.");
```
Comment (kmanamcheri) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java:462]: Improve this error message to state what entitlement should be enabled.

"Provisioning a medium duty data lake cluster is not enabled for this account. Contact Cloudera support to enable CDP_MEDIUM_DUTY_SDX entitlement for the account"

**Pull request merged**